---
title: "Supplementary figures"
author: "Lambda"
date: "`r Sys.Date()`"
output: 
  pdf_document:
    keep_tex: true
bibliography: /home/single_cell_analysis/brain_storm/supp_ref.bib
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      fig.align = "center",
                      message = FALSE,
                      warning = FALSE,
                      results = "hide",
                      echo = FALSE,
                      dev = "png",
                      dpi = 300)
```

```{r}
library(tidyverse)
library(ggalluvial)
library(viridis)
library(OmicsMarkeR)
library(viridis)
library(Seurat)
library(ggpubr)
library(gridExtra)
library(dyno)
name_regex <- "neuron_?10k_v3"
name_use <- "neuron10k_v3"
species <- "Mus musculus"
seus <- readRDS(paste0("./output/", name_use, "/seus_kc_leiden.rds"))
cell_types <- readRDS(paste0("./output/", name_use, "/cell_types.rds"))
markers <- readRDS(paste0("./output/", name_use, "/cluster_all4.rds"))
markers_gsea <- readRDS(paste0("./output/", name_use, "/cluster_gsea.rds"))
n_clusts <- map_dbl(seus, ~ length(unique(.x@meta.data$RNA_snn_res.1)))
kegg_df <- read_csv(paste0("./data/kegg/kegg_df_", str_replace(species, " ", "_"), ".csv"))
clusts <- readRDS(paste0("./output/", name_use, "/clusters_all4.rds"))
clust_assign_summary <- readRDS(paste0("./output/", name_use, "/clust_assign_summary.rds"))
models <- readRDS(paste0("./output/", name_use, "/slingshot_models.rds"))
bcs_inter <- colnames(GetAssayData(seus$kallisto, "data"))
theme_set(theme_bw())
source("./code/plot_comparison.R")
source("./code/gsea.R")
```

## Benchmarking of downstream analyses
We performed detailed analysis of the [10k Brain Cells from an E18 Mouse](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_10k_v3) dataset from 10x for a more thorough comparison between `kallisto bus` and CellRanger in downstream workflows used to generate biologically relevant hypotheses from data. All the downstream analyses are performed with R version 3.5.2.

### Preprocessing
We used the R package `DropletUtils` to remove empty droplets from the kallisto gene count matrix by the inflection point in the total UMI count vs. barcode rank plot. For CellRanger results, we directly used the filtered matrix from CellRanger. After removing the empty droplets, we also removed genes not detected in any remaining barcode in both the kallisto and CellRanger matrices. All downstream analyses were done with the genes and barcodes present in both kallisto and CellRanger matrices unless otherwise specified.

We used Seurat version 3 [@Stuart2018] for basic analysis of this dataset. First, we normalized the data by dividing the UMI count of each gene in each cell by the total UMI counts of that cell, multiplied this number by 10000, added a pseudocount of 1, and applied the natural log. Then we scaled the normalized data so the distribution of the expression of each gene would have mean of 0 and standard deviation of 1. Then we selected 3000 highly variable genes with the `vst` method in Seurat. Then principal component analysis (PCA) was performed on the highly variable genes in the scaled data with the R package `irlba` called by Seurat. The first 40 principal components (PC) were used for tSNE, which was done with the R package `Rtsne` called by Seurat.

### Clustering
Clustering is commonly used in scRNA-seq analysis to classify cells and discover novel cell types. We ran Leiden clustering [@Traag2019] on the gene count matrix generated by `kallisto bus` and CellRanger, with 20 nearest neighbors and resolution 1. Overall, the kallisto an CellRanger results show similar structures in dimension reductions (Supplementary Figure S1). 

### Cluster marker genes
Next, for both kallisto and CellRanger results, we performed differential expression (DE) between each cluster and the rest of the cells to identify and characterize marker genes, with the logistic regression method as described in [@Ntranos2019] and implemented in Seurat, with the normalized (not scaled) data. 

We also performed Spearman and Pearson correlation of UMI counts of top 15 cluster marker genes (by average log fold change) in each barcode present in both the kallisto and CellRanger results. Only the top marker genes are used since those are of interest to experimental validation. Furthermore, including ubiquitously expressed housekeeping genes and lowly expressed genes can introduce misleadingly high correlation even between quite different cells, so the top marker genes give a better sense of how kallisto and CellRanger compare for biological inference. 

We performed gene set enrichment analysis (GSEA) on top 20 cluster marker genes (by average log fold change) for each cluster to facilitate cell type annotation. GSEA was performed using the R package `EGSEA` [@Alhamdoosh2017], and the gene sets used are KEGG pathways. Since KEGG uses Entrez gene ID for mice, only genes with Entrez gene IDs were included in GSEA.

### Pseudotime
Since this dataset is from an E18 mouse embryo, pseudotime analysis is relevant. We used SingleR [@Aran2019] to more systematically annotate cell types based on correlation of gene expression profiles of cells to reference bulk RNA-seq data of isolated cell types of the mouse brain from [@Benayoun2019]. The main neuronal cell types are neuronal precursor cells (NPC), neurons, active neuronal stem cells (aNSC), and quiescent neuronal stem cells (qNSC). Only the neuronal cell types were used for pseudotime analysis, since other cell types, such as erythrocytes and endothelial cells may be mistaken as a divergent branch of the trajectory distant in pseudotime. We set the neurons to be the end points of the trajectory. Pseudotime analysis was done with the `slingshot` [@Street2018] via the Docker container from `dyno` [@Saelens2019].

To more quantitatively compare pseudotime inference for kallisto and CellRanger results, we applied a number of metrics developed by [@Saelens2019] to evaluate pseudotime methods. These metrics should give scores with absolute values between 0 and 1; having a value closer to 1 means more similarity. The first metric used is correlation in geodeisic distance of cells on the trajectory, i.e. relative position of cells. This compares cell ordering between kallisto and CellRanger results. We also compared how branches and milestones (assignment of cells to stages) assigned by pseudotime analysis for kallisto and CellRanger results overlap. Genes predictive of cells' positions on the trajectory may be driving or influenced by cell differentiation, so can be biologically relevant. Thus we computed correlation of feature importance between kallisto and CellRanger pseudotime results, weighted so that more predictive genes contribute more to the correlation. Scores from these metrics from comparing kallisto and CellRanger results are generally close to 1, indicating similarity in the pseudotime results:

Metric | Value
-------|--------
Geodeisic distance correlation | 0.874
Branch overlap | 0.845
Milestone overlap | 0.794
Feature importance weighted correlation | 0.905

### DE genes between kallisto and CellRanger results
To investigate which genes are impacted by choosing kallisto as opposed to CellRanger, we ran DE between the matrices from the two workflows. First, we concatenated the two matrices using the intersection of genes. Then we normalized the concatenated matrix with Seurat. DE was performed with logistic regression as when finding cluster marker genes. Next we performed GSEA with the R package `EGSEA` on all marker genes with adjusted p-value less than 0.05 to identify types of genes more likely to be affected by the different workflows. Again, KEGG pathways were used as gene sets. 

### Species mixing
To explore the performance of kallisto in mixed species data, we compared the number and proportion of human and mouse UMIs in each cell in the [10k 1:1 Mixture of Fresh Frozen Human (HEK293T) and Mouse (NIH3T3) Cells](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/hgmm_10k_v3) dataset. Human and mouse genes were identified with their Ensembl gene ID. In Supplementary Figure S6A, the total number of UMIs mapped to human and mouse genes in each barcode was calculated in the unfiltered matrices. In Supplementary Figure S6B,C, only barcodes present in both kallisto and CellRanger unfiltered matrices were used. 

# Figures
```{r fig4.1, fig.height=10, fig.width=8}
fig1a <- imap(seus, ~ ElbowPlot(.x, ndims = 60) + ggtitle(.y) + theme_bw()) %>% 
  ggarrange(plotlist = ., ncol = 2)
fig1b <- DimPlot_list(seus, pt.size = 0.3)
fig1c <- DimPlot_list(seus, "tsne", pt.size = 0.3)
ggarrange(fig1a, fig1b, fig1c, labels = LETTERS[1:3], ncol = 1, nrow = 3)
```

**Supplementary Figure 1.** (A) Elbow plot of standard deviation explained by each principal component (PC) of gene count matrix from kallisto and CellRanger. (B) Cell embedding in the first 2 PCs colored by cluster. (C) Cell embedding in tSNE colored by cluster.

```{r}
clusts_alluvial <- clusts %>% 
  group_by_if(is.factor) %>% 
  dplyr::count()
```

```{r fig4.2, fig.width=8, fig.height=7}
fig2a <- ggplot(clusts_alluvial, aes(y = n, axis1 = kallisto, axis2 = cellranger)) +
  geom_flow(aes(fill = kallisto, color = kallisto), size = 0.3, width = 1/6) +
  scale_x_discrete(limits = c("kallisto", "cellranger"), expand = c(0,0)) +
  geom_stratum(width = 1/6) +
  geom_text(stat = "stratum", label.strata = TRUE) +
  labs(y = "Number of cells") +
  scale_fill_viridis_d(name = "kallisto cluster", option = "E") +
  scale_color_viridis_d(name = "kallisto cluster", option = "E") +
  scale_y_continuous(expand = c(0,0)) +
  theme(legend.position = "none",
        panel.grid = element_blank())
fig2b <- ggplot(clust_assign_summary, aes(kallisto, cluster, fill = jaccard)) +
  geom_tile() +
  scale_fill_viridis_c() +
  coord_equal() +
  theme(panel.background = element_rect(fill = viridis(256)[1]),
        panel.grid = element_blank()) +
  labs(x = "kallisto cluster", y = "CellRanger cluster") +
  scale_x_discrete(expand = c(0,0)) +
  scale_y_discrete(expand = c(0,0)) +
  theme(legend.position = "top")
ggarrange(fig2a, fig2b, nrow = 1, ncol = 2, labels = LETTERS[1:2])
```

**Supplementary Figure 2.** (A) Number of cells assigned to each cluster in kallisto results and CellRanger results. (B) Jaccard indices between each kallisto cluster and each CellRanger cluster.

```{r fig4.3, fig.width=12, fig.height=14}
fig3a <- plot_n_markers(markers, 0.75, ncol = 2, nrow = 1, positive_only = TRUE, nudge_y = 30) +
  labs(y = "Number of marker genes")
fig3b <- ggplot(markers, aes(cluster, avg_logFC)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1) +
  facet_wrap(~ method, ncol = 2) +
  labs(y = "Log fold change")
fig3c <- markers %>% 
  top_n(20, avg_logFC) %>% 
  gsea_bubble(kegg_df, markers_gsea, species, 4)
ggarrange(fig3a, fig3b, fig3c, ncol = 1, nrow = 3, labels = LETTERS[1:3], heights = c(1,1,4))
```

**Supplementary Figure 3.** (A) Number of marker genes with log fold change of at least 0.75 and adjusted p < 0.05 in each cluster. (B) Log fold change of marker genes in each cluster. (C) Gene set enrichment of top 20 marker genes (by log fold change) in each cluster.

```{r}
topn <- 15
top_markers_list <- markers %>% 
  filter(rank_logFC <= topn, avg_logFC > 0.75)
top_markers_list <- split(top_markers_list$gene, top_markers_list$method)
top_markers_inter <- intersect(top_markers_list[[1]], top_markers_list[[2]])
```

```{r}
cor_use <- map(seus, ~ as.matrix(GetAssayData(.x, "counts")[top_markers_inter, ]))
cors_spearman <- map_dbl(bcs_inter, ~ cor(cor_use$kallisto[, .x], cor_use$cellranger[, .x], 
                                          method = "spearman"))
cors_pearson <- map_dbl(bcs_inter, ~ cor(cor_use$kallisto[, .x], cor_use$cellranger[, .x], 
                                          method = "pearson"))
# Color by cluster in kallisto
cors <- tibble(spearman = cors_spearman,
               pearson = cors_pearson,
               clust_use = seus$kallisto$seurat_clusters) %>% 
  gather("coefficient", "value", -clust_use)
```

```{r fig4.4, fig.height=8, fig.width=8}
fig4a <- ggplot(cors, aes(value)) +
  geom_histogram(bins = 100) +
  labs(x = "Correlation coefficient", y = "count") +
  #scale_fill_viridis_d(name = "kallisto cluster", option = "E") +
  facet_wrap(~ coefficient, ncol = 1, strip.position = "right") +
  theme(legend.position = "top")
fig4b <- ggplot(cors, aes(clust_use, value)) +
  geom_violin() +
  geom_jitter(size = 0.1, alpha = 0.1) +
  facet_wrap(~ coefficient, ncol = 1, strip.position = "right") +
  labs(x = "kallisto cluster", y = "Correlation coefficient")
ggarrange(fig4a, fig4b, ncol = 1, nrow = 2, labels = LETTERS[1:2])
```

**Supplementary Figure 4.** (A) Histograms of Spearman and Pearson correlation coefficients between barcodes from kallisto and the same barcodes from CellRanger, with top 15 marker genes (by log fold change) of each cluster. (B) Spearman and Pearson correlation coefficients, as in (A), for cells in each kallisto cluster. Here cluster 16 is erythrocytes, while most other cells are neuronal precursor cells.

```{r}
for (i in names(seus)) {
  Idents(seus[[i]]) <- "label"
}
neurons_inter <- models$kallisto$cell_ids
seus_neuron <- map(seus, ~subset(.x, cells = neurons_inter))
```

```{r fig4.5, fig.width=9, fig.height=6}
fig5a <- pmap(list(models[c("kallisto", "cellranger")], seus_neuron, names(seus_neuron)), 
             ~ plot_dimred(..1, grouping = ..2$label, alpha_cells = 0.5) +
               ggtitle(..3)) %>% 
  ggarrange(plotlist = ., ncol = 2, nrow = 1, common.legend = TRUE, legend = "bottom")
fig5b <- pmap(list(models[c("kallisto", "cellranger")], seus_neuron, names(seus_neuron)), 
             ~ plot_dimred(..1, color_cells = "pseudotime", alpha_cells = 0.5) +
               ggtitle(..3)) %>% 
  ggarrange(plotlist = ., ncol = 2, nrow = 1, common.legend = TRUE, legend = "bottom")
ggarrange(fig5a, fig5b, ncol = 1, nrow = 2, labels = LETTERS[1:2])
```

**Supplementary Figure 5.** (A) Lineage inference with `slingshot` [@Street2018], projected to the first 2 principal components, with cells colored by cell type inferred by `SingleR` [@Aran2019]. The reference used for `SingleR` is from [@Benayoun2019]. aNSCs stands for active neuronal stem cells. NPCs stands for neuronal precursor cells. qNSCs stands for quiescent neuronal stem cells. (B) Coloring by pseudotime values from `slingshot`.

```{r}
cell_species <- readRDS("./output/hgmm10k_v3/cell_species_full.rds")
cell_species_compare <- readRDS("./output/hgmm10k_v3/cell_species_compare_full.rds")
```

```{r fig4.6, fig.width=8, fig.height=8}
fig6a <- ggplot(cell_species, aes(n_human_umi, n_mouse_umi, color = species)) +
  geom_point(size = 0.5, alpha = 0.5) +
  facet_wrap(~ method, ncol = 2) +
  labs(x = "Number of human UMIs", y = "Number of mouse UMIs")
fig6b <- ggplot(cell_species_compare, aes(tot_umi.x, tot_umi.y)) +
  geom_point(size = 0.5, alpha = 0.1) +
  geom_abline(slope = 1, intercept = 0, color = "red", size = 0.3) +
  annotate("text", x = 1e3, y = 1e4, label = "y = x", color = "red") +
  labs(x = "kallisto", y = "CellRanger", title = "Total counts per barcode") +
  coord_equal() +
  scale_x_log10() +
  scale_y_log10()
fig6c <- ggplot(cell_species_compare, aes(prop_human.x, prop_human.y, 
                                          alpha = avg_counts)) +
  geom_point(size = 0.5) +
  geom_abline(slope = 1, intercept = 0, color = "red", size = 0.3) +
  annotate("text", x = 0.5, y = 0.75, label = "y = x", color = "red") +
  labs(x = "kallisto", y = "CellRanger", title = "Proportion of human UMIs") +
  coord_equal() +
  scale_alpha_continuous(range = c(1.925e-3, 1), name = "Average total \nUMI counts") +
  theme(legend.position = c(0.83, 0.22), 
        legend.background = element_blank())
ggarrange(fig6a,
          ggarrange(fig6b, fig6c, nrow = 1, ncol = 2, labels = c("B", "C")),
          nrow = 2, ncol = 1,
          labels = "A")
```

**Supplementary Figure 6.** (A) Barnyard plot for hgmm10k dataset from 10x processed with CellRanger and kallisto. (B) Consistency of total counts per barcode between kallisto results and CellRanger results (before removing empty droplets). (C) Proportion of UMIs from human genes in kallisto and CellRanger results (before removing empty droplets). Opacity of the points is proportional to the average of total UMI counts in kallisto results and CellRanger results.

# Supplementary references
