{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "bustools_mash_merge.ipynb",
      "provenance": [],
      "authorship_tag": "ABX9TyOryrRQOmQmbgoxG22bHv47",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/pachterlab/MBGBLHGP_2019/blob/master/notebooks/merge/bustools_mash_merge.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7LrGhLCQUf6W",
        "outputId": "6d35674f-211f-4150-989d-a435f295a209",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 448
        }
      },
      "source": [
        "# install kallisto and bustools and seqkit\n",
        "!git clone https://github.com/pachterlab/MBGBLHGP_2019.git\n",
        "!cp MBGBLHGP_2019/notebooks/merge/kallisto /usr/local/bin/\n",
        "!cp MBGBLHGP_2019/notebooks/merge/bustools /usr/local/bin/\n",
        "!wget https://github.com/shenwei356/seqkit/releases/download/v0.13.2/seqkit_linux_amd64.tar.gz && tar -xvf seqkit_linux_amd64.tar.gz && cp seqkit /usr/local/bin/"
      ],
      "execution_count": 21,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'MBGBLHGP_2019'...\n",
            "remote: Enumerating objects: 18, done.\u001b[K\n",
            "remote: Counting objects: 100% (18/18), done.\u001b[K\n",
            "remote: Compressing objects: 100% (12/12), done.\u001b[K\n",
            "remote: Total 299 (delta 3), reused 17 (delta 2), pack-reused 281\n",
            "Receiving objects: 100% (299/299), 87.19 MiB | 39.11 MiB/s, done.\n",
            "Resolving deltas: 100% (114/114), done.\n",
            "--2020-10-15 18:48:57--  https://github.com/shenwei356/seqkit/releases/download/v0.13.2/seqkit_linux_amd64.tar.gz\n",
            "Resolving github.com (github.com)... 192.30.255.113\n",
            "Connecting to github.com (github.com)|192.30.255.113|:443... connected.\n",
            "HTTP request sent, awaiting response... 302 Found\n",
            "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/52715040/00fa4280-c509-11ea-8f41-bb3e8dd6f501?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20201015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20201015T184857Z&X-Amz-Expires=300&X-Amz-Signature=24ab061de30f615bbaaab2649cf78d268ee28803cad7ecf82c4bcd99adea9b8e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
            "--2020-10-15 18:48:57--  https://github-production-release-asset-2e65be.s3.amazonaws.com/52715040/00fa4280-c509-11ea-8f41-bb3e8dd6f501?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20201015%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20201015T184857Z&X-Amz-Expires=300&X-Amz-Signature=24ab061de30f615bbaaab2649cf78d268ee28803cad7ecf82c4bcd99adea9b8e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=52715040&response-content-disposition=attachment%3B%20filename%3Dseqkit_linux_amd64.tar.gz&response-content-type=application%2Foctet-stream\n",
            "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.217.96.172\n",
            "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.217.96.172|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 7486046 (7.1M) [application/octet-stream]\n",
            "Saving to: ‘seqkit_linux_amd64.tar.gz.1’\n",
            "\n",
            "seqkit_linux_amd64. 100%[===================>]   7.14M  13.6MB/s    in 0.5s    \n",
            "\n",
            "2020-10-15 18:48:58 (13.6 MB/s) - ‘seqkit_linux_amd64.tar.gz.1’ saved [7486046/7486046]\n",
            "\n",
            "seqkit\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nQd5MhOVcTT_"
      },
      "source": [
        "!mkdir -p split full ref/split ref/full"
      ],
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aiWXEt32cBk7",
        "outputId": "808ec6df-c494-4d80-cc07-700da706e843",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 274
        }
      },
      "source": [
        "# download the transcriptome\n",
        "!wget ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2020-10-15 18:06:35--  ftp://ftp.ensembl.org/pub/release-101/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
            "           => ‘Homo_sapiens.GRCh38.cdna.all.fa.gz’\n",
            "Resolving ftp.ensembl.org (ftp.ensembl.org)... 193.62.193.139\n",
            "Connecting to ftp.ensembl.org (ftp.ensembl.org)|193.62.193.139|:21... connected.\n",
            "Logging in as anonymous ... Logged in!\n",
            "==> SYST ... done.    ==> PWD ... done.\n",
            "==> TYPE I ... done.  ==> CWD (1) /pub/release-101/fasta/homo_sapiens/cdna ... done.\n",
            "==> SIZE Homo_sapiens.GRCh38.cdna.all.fa.gz ... 69368550\n",
            "==> PASV ... done.    ==> RETR Homo_sapiens.GRCh38.cdna.all.fa.gz ... done.\n",
            "Length: 69368550 (66M) (unauthoritative)\n",
            "\n",
            "Homo_sapiens.GRCh38 100%[===================>]  66.15M  17.3MB/s    in 3.8s    \n",
            "\n",
            "2020-10-15 18:06:41 (17.3 MB/s) - ‘Homo_sapiens.GRCh38.cdna.all.fa.gz’ saved [69368550]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "o4gm7TSDgHG8"
      },
      "source": [
        "## Full transcriptome"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "nKIa8F6TdH4A",
        "outputId": "4a6deb43-1850-4e77-9782-fd6376384e88",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 277
        }
      },
      "source": [
        "!kallisto index -i index.idx Homo_sapiens.GRCh38.cdna.all.fa.gz"
      ],
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 1484 target sequences\n",
            "[build] warning: replaced 100005 non-ACGUT characters in the input sequence\n",
            "        with pseudorandom nucleotides\n",
            "[build] counting k-mers ... tcmalloc: large alloc 1610612736 bytes == 0x55eb8cd8a000 @  0x7f683f604887 0x55eb12a1117d 0x55eb12a19ff9 0x55eb129e7387 0x7f683e484b97 0x55eb129ec26a\n",
            "tcmalloc: large alloc 3221225472 bytes == 0x55ebecd8a000 @  0x7f683f604887 0x55eb12a1117d 0x55eb12a19ff9 0x55eb129e7387 0x7f683e484b97 0x55eb129ec26a\n",
            "done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 1155300 contigs and contains 111187137 k-mers \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uYxLIuCQgM8i"
      },
      "source": [
        "!zcat Homo_sapiens.GRCh38.cdna.all.fa.gz | awk '(NR-1)%2==0{print $1, $4}' OFS=\"\\t\"| tr -d \"gene:\" | tr -d \">\" > ref/full/t2g.txt"
      ],
      "execution_count": 16,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "BBekqijZgKSP"
      },
      "source": [
        "## Split transcriptome"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "8zWVcJiicWkl",
        "outputId": "0f670adb-7cf5-46f7-e2df-f56526a0cb10",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 206
        }
      },
      "source": [
        "!seqkit split -p 8 Homo_sapiens.GRCh38.cdna.all.fa.gz"
      ],
      "execution_count": 10,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[INFO]\u001b[0m split into 8 parts\n",
            "[INFO]\u001b[0m read sequences ...\n",
            "[INFO]\u001b[0m read 191887 sequences\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_001.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_002.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_003.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_004.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_005.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_006.fa.gz\n",
            "[INFO]\u001b[0m write 23986 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_007.fa.gz\n",
            "[INFO]\u001b[0m write 23985 sequences to file: Homo_sapiens.GRCh38.cdna.all.fa.gz.split/Homo_sapiens.GRCh38.cdna.all.part_008.fa.gz\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "za7Hc8KCc4SG",
        "outputId": "9091503e-2e03-4c40-ed5c-76dd9c332bdc",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "!cd Homo_sapiens.GRCh38.cdna.all.fa.gz.split/ && for f in *.gz; do kallisto index -i $(echo $f |  cut -d\".\" -f5).idx $f; done"
      ],
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_001.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 188 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 120678 contigs and contains 15497713 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_002.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 185 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 131192 contigs and contains 15722480 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_003.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 185 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 137535 contigs and contains 13548803 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_004.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 189 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 130967 contigs and contains 15589113 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_005.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 228 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 120085 contigs and contains 16136097 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_006.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 216 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 123888 contigs and contains 16800168 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_007.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 193 target sequences\n",
            "[build] warning: replaced 100005 non-ACGUT characters in the input sequence\n",
            "        with pseudorandom nucleotides\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 95127 contigs and contains 10905242 k-mers \n",
            "\n",
            "\n",
            "[build] loading fasta file Homo_sapiens.GRCh38.cdna.all.part_008.fa.gz\n",
            "[build] k-mer length: 31\n",
            "[build] warning: clipped off poly-A tail (longer than 10)\n",
            "        from 100 target sequences\n",
            "[build] counting k-mers ... done.\n",
            "[build] building target de Bruijn graph ...  done \n",
            "[build] creating equivalence classes ...  done\n",
            "[build] target de Bruijn graph has 178696 contigs and contains 13232906 k-mers \n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RQVY4FElenBM"
      },
      "source": [
        "!mv index.idx ref/full/\n",
        "!mv Homo_sapiens.GRCh38.cdna.all.fa.gz ref/full\n",
        "!mv Homo_sapiens.GRCh38.cdna.all.fa.gz.split/ ref/split"
      ],
      "execution_count": 19,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "f_L_u_Fel1OG"
      },
      "source": [
        "## Full alignment"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "JDAl_pTnmFyH",
        "outputId": "5299cb0e-79b2-4eba-976b-629762b5a681",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 277
        }
      },
      "source": [
        "# kallisto bus on full guy\n",
        "!time kallisto bus -x 10xv2 -i ref/full/index.idx -o full/o -t 2 \\\n",
        "/content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz \\\n",
        "/content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz"
      ],
      "execution_count": 22,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 191,887\n",
            "[index] number of k-mers: 111,187,137\n",
            "tcmalloc: large alloc 3221225472 bytes == 0x55cc8470a000 @  0x7f5e5edfc887 0x55cc82690fe2 0x55cc82689571 0x55cc8265d509 0x7f5e5dc7cb97 0x55cc8266226a\n",
            "[index] number of equivalence classes: 778,047\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 606,419 reads pseudoaligned\n",
            "\n",
            "real\t0m56.477s\n",
            "user\t0m27.145s\n",
            "sys\t0m3.464s\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "61y1Q_Yne0xM",
        "outputId": "8d39f031-5f68-4a79-e835-0ab1182261d4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 191
        }
      },
      "source": [
        "!bustools sort -o full/o/s.bus full/o/output.bus\n",
        "!bustools correct -o full/o/cs.bus -w /content/MBGBLHGP_2019/notebooks/merge/10xv2_whitelist.txt full/o/s.bus\n",
        "!bustools sort -o full/o/scs.bus full/o/cs.bus\n",
        "!mkdir full/o/count\n",
        "!bustools count -o full/o/count/count -g ref/full/t2g.txt -e full/o/matrix.ec -t full/o/transcripts.txt --genecounts full/o/scs.bus"
      ],
      "execution_count": 23,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x556b35402000 @  0x7f5872292887 0x556b335d13fd 0x556b33592f4a 0x7f5871112b97 0x556b3359407a\n",
            "Read in 606419 BUS records\n",
            "Found 737280 barcodes in the whitelist\n",
            "Processed 335382 BUS records\n",
            "In whitelist = 322684\n",
            "Corrected    = 3402\n",
            "Uncorrected  = 9296\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x555f224d2000 @  0x7f34b993b887 0x555f20c6a3fd 0x555f20c2bf4a 0x7f34b87bbb97 0x555f20c2d07a\n",
            "Read in 326086 BUS records\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Cr1ugdwzpwvs"
      },
      "source": [
        "## Split alignment"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4WF6F-3Ml9Nx"
      },
      "source": [
        "!mkdir split/o1 split/o2 split/o3 split/o4 split/o5 split/o6 split/o7 split/o8"
      ],
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "fIp88RvYp_lj",
        "outputId": "cb72bdd0-9fae-4e8f-b106-74eec38cedbe",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        }
      },
      "source": [
        "!for d in ./split/o*; do kallisto bus -k -n -x 10xv2 -o $d \\\n",
        "-i ref/split/Homo_sapiens.GRCh38.cdna.all.fa.gz.split/part_00$(echo $d | cut -c10).idx \\\n",
        "/content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz \\\n",
        "/content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz; done"
      ],
      "execution_count": 35,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 15,497,713\n",
            "[index] number of equivalence classes: 88,114\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,591,752 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 15,722,480\n",
            "[index] number of equivalence classes: 96,419\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,828,298 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 13,548,803\n",
            "[index] number of equivalence classes: 98,980\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,896,919 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 15,589,113\n",
            "[index] number of equivalence classes: 94,000\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,791,591 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 16,136,097\n",
            "[index] number of equivalence classes: 90,031\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,642,981 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 16,800,168\n",
            "[index] number of equivalence classes: 89,274\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,698,697 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,986\n",
            "[index] number of k-mers: 10,905,242\n",
            "[index] number of equivalence classes: 74,514\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 2,319,075 kmers pseudoaligned\n",
            "\n",
            "[index] k-mer length: 31\n",
            "[index] number of targets: 23,985\n",
            "[index] number of k-mers: 13,232,906\n",
            "[index] number of equivalence classes: 88,079\n",
            "[quant] will process sample 1: /content/MBGBLHGP_2019/notebooks/merge/fastqs/r1.fastq.gz\n",
            "                               /content/MBGBLHGP_2019/notebooks/merge/fastqs/r2.fastq.gz\n",
            "[quant] finding pseudoalignments for the reads ... done\n",
            "[quant] processed 1,000,000 reads, 3,483,865 kmers pseudoaligned\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "7suFTSfBtoiF",
        "outputId": "43e7a36e-c1f2-42bc-fcea-114d465e1fbd",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 311
        }
      },
      "source": [
        "!for d in ./split/o*; do bustools sort --flags -o $d/flagsort.bus $d/output.bus; done"
      ],
      "execution_count": 36,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x559f236d8000 @  0x7f489df4e887 0x559f215653fd 0x559f21526f4a 0x7f489cdceb97 0x559f2152807a\n",
            "Read in 5183504 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x558527632000 @  0x7fb4a3cb9887 0x5585252ab3fd 0x55852526cf4a 0x7fb4a2b39b97 0x55852526e07a\n",
            "Read in 5656596 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x5599f7f10000 @  0x7f5b6671a887 0x5599f5a3b3fd 0x5599f59fcf4a 0x7f5b6559ab97 0x5599f59fe07a\n",
            "Read in 5793838 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x55e6aa37c000 @  0x7f80fbdef887 0x55e6a8e443fd 0x55e6a8e05f4a 0x7f80fac6fb97 0x55e6a8e0707a\n",
            "Read in 5583182 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x5581f04a4000 @  0x7f20034bd887 0x5581ee7de3fd 0x5581ee79ff4a 0x7f200233db97 0x5581ee7a107a\n",
            "Read in 5285962 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x55e0c02f8000 @  0x7f2620c1d887 0x55e0bf6dd3fd 0x55e0bf69ef4a 0x7f261fa9db97 0x55e0bf6a007a\n",
            "Read in 5397394 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x55d25a7da000 @  0x7f067241e887 0x55d257ef63fd 0x55d257eb7f4a 0x7f067129eb97 0x55d257eb907a\n",
            "Read in 4638150 BUS records\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x55dbe7aea000 @  0x7f188aa78887 0x55dbe58b43fd 0x55dbe5875f4a 0x7f18898f8b97 0x55dbe587707a\n",
            "Read in 6967730 BUS records\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TvHVAfP6ts8h"
      },
      "source": [
        "!for d in ./split/o*; do mv $d/flagsort.bus $d/output.bus; done"
      ],
      "execution_count": 37,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "aBKKQk-luPWC",
        "outputId": "8c61ba6e-8095-4870-cc25-e700cdbef772",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 103
        }
      },
      "source": [
        "!bustools mash -o ./split/mashed ./split/o*"
      ],
      "execution_count": 38,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[info] parsed output.bus files\n",
            "[info] parsed transcripts.txt\n",
            "[info] parsed matrix.ec files\n",
            "bus records read:    44067739\n",
            "bus records written: 44067739\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WOsw39vIurVr",
        "outputId": "4e17b85a-6860-4379-858c-8bd4b4c5613b",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 71
        }
      },
      "source": [
        "!bustools sort --flags -o ./split/mashed/smashed.bus ./split/mashed/mashed.bus"
      ],
      "execution_count": 40,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x55c3fb9fe000 @  0x7f6d2d00a887 0x55c3fa2b73fd 0x55c3fa278f4a 0x7f6d2be8ab97 0x55c3fa27a07a\n",
            "Read in 44067739 BUS records\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "MbagnDm5uWds",
        "outputId": "47205989-e74c-4ccb-a99c-2b1d9a81df94",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 103
        }
      },
      "source": [
        "!bustools merge -o ./split/mashed/merged.bus -e ./split/mashed/matrix.ec -t ./split/mashed/transcripts.txt ./split/mashed/smashed.bus"
      ],
      "execution_count": 42,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[info] parsed transcripts.txt\n",
            "[info] parsed matrix.ec\n",
            "bus records read:    44067732\n",
            "bus records written: 310361\n",
            "bus records lost:    483981\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "tgpRfvhPuceJ"
      },
      "source": [
        "!mv merged.ec ./split/mashed/"
      ],
      "execution_count": 43,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ofTcVUsvxSQ6",
        "outputId": "58173b1d-033a-488c-ca4a-a6b0042e7cf1",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 191
        }
      },
      "source": [
        "!bustools sort -o ./split/mashed/s.bus ./split/mashed/merged.bus\n",
        "!bustools correct -o ./split/mashed/cs.bus -w /content/MBGBLHGP_2019/notebooks/merge/10xv2_whitelist.txt ./split/mashed/s.bus\n",
        "!bustools sort -o ./split/mashed/scs.bus ./split/mashed/cs.bus\n",
        "!mkdir ./split/mashed/count\n",
        "!bustools count -o ./split/mashed/count/count -g ref/full/t2g.txt -e ./split/mashed/merged.ec -t ./split/mashed/transcripts.txt --genecounts ./split/mashed/scs.bus"
      ],
      "execution_count": 44,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "tcmalloc: large alloc 4294967296 bytes == 0x561894f16000 @  0x7fd0ac0cd887 0x56189295c3fd 0x56189291df4a 0x7fd0aaf4db97 0x56189291f07a\n",
            "Read in 310361 BUS records\n",
            "Found 737280 barcodes in the whitelist\n",
            "Processed 310361 BUS records\n",
            "In whitelist = 300464\n",
            "Corrected    = 2030\n",
            "Uncorrected  = 7867\n",
            "tcmalloc: large alloc 4294967296 bytes == 0x562f777dc000 @  0x7fd144257887 0x562f763373fd 0x562f762f8f4a 0x7fd1430d7b97 0x562f762fa07a\n",
            "Read in 302494 BUS records\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kEUv-PXYxoia"
      },
      "source": [
        "# Comparisons"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Un1ke1xzxoOF"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Eo2i-E0cxgt4"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}