# Code for reproducing the figures constituting Supplementary Figure 6 and Supplementary Figure 7 from the paper "Modular and efficient pre-processing of single-cell RNA-seq"

These figures are from an in depth analysis of the 10x [10k brain cells from an E18 mice](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_10k_v3) (Figure 6) and the species mixing analysis for the 10x [10k hgmm dataset](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/hgmm_10k_v3) (Figure 7). The code in this directory is dependent on the code for making Figure 1, which is used to generate the gene count matrices for the datasets.

All notebooks used for the analyses are in the directory `analysis`. Notebooks for the brain 10k dataset are in the subdirectory `neuron10k_v3`. These notebooks include code used to produce figures not included in the paper, but that were useful for data analysis, as well as some relevant analyses and comments. The figures themselves were generated with `supp_fig.Rmd`.

R scripts with functions used in the notebooks are in the directory `code`. In addition, in this directory, `ensembl_kegg_ref.R` was used to download Ensembl gene annotation and KEGG pathways for all the 20 benchmark panel datasets used in this study, and `gsea_bar_full.R` was used to run differential expression between kallisto and Cell Ranger and to perform gene set enrichment analysis (GSEA) for benchmark panel datasets. The results from `gsea_bar_full.R` are not included in Supplementary Figure 6; they constitute panel (H) in the benchmark panel figures from the Supplement Figures.
