# Code for reproducing results of the paper "Modular and efficient pre-processing of single-cell RNA-seq"
These figures are from an in depth analysis of the 10x [10k brain cells from an E18 mice](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/neuron_10k_v3) and the species mixing analysis for the 10x [10k hgmm dataset](https://support.10xgenomics.com/single-cell-gene-expression/datasets/3.0.0/hgmm_10k_v3). The code in this directory on its own can't be run from beginning to end, since the code required to generate the gene count matrices of the datasets is not included. The gene count matrices were made from code for Figure 3.

All notebooks used for the analyses are in the directory `analysis`. Notebooks for the brain 10k dataset are in the subdirectory `neuron10k_v3`. These notebooks include code used to produce figures not included in the paper, but were useful for data analysis, as well as some relevant analyses and comments that did not make it to the paper. The figures themselves were generated with `supp_fig.Rmd`.

R scripts with functions used in the notebooks are in the directory `code`. In addition, in this directory, `ensembl_kegg_ref.R` was used to download Ensembl gene annotation and KEGG pathways for all the 21 datasets used in this study, and `gsea_bar_full.R` was used to run differential expression between kallisto and CellRanger and do gene set enrichment analysis for all 21 datasets. The results from `gsea_bar_full.R` is not included in Supplementary Figure 4; it's in one of the 8 panel summary figures in the Supplement. The script `alevin_mtx.R` was used to read Salmon Alevin binary output into R as a sparse matrix and write it as a `mtx` file.
